### ISSUE TITLE
Power management

### ISSUE THREAD

**Issue by omariom** [dotnet/BenchmarkDotNet#68]
My ultrabook's CPU frequency goes up and down accommodating to the load.
That's because it is in Balanced power mode. To keep it upmost I set it to High performance manually.

It would be convenient if BenchmarkDotNet  did that for us.

This is what I use in my microbechmarking tool for LinqPad.
I found it in Vance Morrison's MeasureIt.

``` C#
using System;
using System.Runtime.InteropServices;
using System.Diagnostics;

/// <summary>
/// PowerManagement allows you to access the funtionality of the Control Panel -> Power Options
/// dialog in windows.  (Currently we only use VISTA APIs). 
/// </summary>
public static unsafe class PowerManagment
{
    public static Guid HighPerformance = new Guid(0x8c5e7fda, 0xe8bf, 0x4a96, 0x9a, 0x85, 0xa6, 0xe2, 0x3a, 0x8c, 0x63, 0x5c);
    public static Guid Balenced        = new Guid(0x381b4222, 0xf694, 0x41f0, 0x96, 0x85, 0xff, 0x5b, 0xb2, 0x60, 0xdf, 0x2e);
    public static Guid PowerSaver      = new Guid(0xa1841308, 0x3541, 0x4fab, 0xbc, 0x81, 0xf7, 0x15, 0x56, 0xf2, 0x0b, 0x4a);

    public static Guid CurrentPolicy
    {
        get
        {
            Guid* retPolicy = null;
            Guid ret = Guid.Empty;
            try
            {
                int callRet = PowerGetActiveScheme(IntPtr.Zero, ref retPolicy);
                if (callRet == 0)
                {
                    ret = *retPolicy;
                    Marshal.FreeHGlobal((IntPtr)retPolicy);
                }
            }
            catch (Exception) { }
            return ret;
        }
    }
    public static bool Set(Guid newPolicy)
    {
        try
        {
            return PowerSetActiveScheme(IntPtr.Zero, ref newPolicy) == 0;
        }
        catch (Exception) { }
        return false;
    }

    #region private 
        [DllImport("powrprof.dll")]
    private static extern int PowerGetActiveScheme(IntPtr ReservedZero, ref Guid* policyGuidRet);

    [DllImport("powrprof.dll")]
    private static extern int PowerSetActiveScheme(IntPtr ReservedZero, ref Guid policyGuid);
    #endregion
}
```

**Comment by AndreyAkinshin** [comment]
It is a very good idea, I like it. Another good links about the subject:
https://msdn.microsoft.com/en-us/library/system.windows.forms.systeminformation.aspx
https://msdn.microsoft.com/en-us/library/hxkc1kwd.aspx
http://thoughtso.org/wp-content/uploads/2007/11/cpower.txt
http://stackoverflow.com/questions/16023978/how-to-use-power-management-functions-powerenuimerate-to-get-power-settings
http://www.fsmpi.uni-bayreuth.de/~dun3/archives/programmatically-change-power-options-using-cshar/519.html

**Comment by mattwarren** [comment]
Agree, that's a great idea, I've used Vance Morrison's MeasureIt before, but I never realised that it forced the power mode like that.

**Comment by mattwarren** [comment]
Some more related info (from a comment on Scott Hanselman's Blog Post ["Benchmarking .NET code"](http://www.hanselman.com/blog/CommentView.aspx?guid=31851866-03d3-4a11-90a0-a50597dca7e2#commentstart)):

> We benchmark our code all the time. For serialization we have done this: [SERBENCH](https://github.com/aumcode/serbench) here are some charts: [Typical Person Serialization](http://aumcode.github.io/serbench/Specimens_Typical_Person/web/overview-charts.htm). 
What is interesting is that the most benchmarks provided by software makers are very skewed, i.e. serializers depend greatly on payload type, as our test indicated. Some things taken for granted like "Protobuff" is the fastest are not true in some test cases you see unexpected results. That's is why it is important to execute the same suite of test cases against different vendors. Sometimes benchmarking something is more work than writing the component, for example SpinLock class has many surprises.

> **Another thing that 90% of people I talk to disregard: SpeedStep. Try to disable it via control panel and dont be surprised that you get 25% perf boost EVEN IF you ran your tests for minutes. SpeedStep does magic tricks, it lowers you CPU clock even if the CPU is swamped, but not on all cores. Running tests from within VS(with debugger attached) can slow things down by good 45-70%.**
 
> So I would say:

> 1. Test different work patterns
1. Build with Optimize
1. Run standalone (CTRL+f5)
1. Disable "Power Saver"/SpeedStep
1. Run for a reasonable time (not 2 seconds)
1. For multi-threaded perf dont forget to set GS SERVER MODE, otherwise your "new" will be 10-40% slower than it could
1. Profilers sometimes lie! They don't guarantee nothing. Great tools but sometimes they don't show whats happening and may be source of Heisenbugs

**Comment by adamsitnik** [comment]
there was no progress for 8 months, so I made it up for grabs again

**Comment by MarekM25** [comment]
Hi,
I am new in Benchmark.NET. @adamsitnik, @AndreyAkinshin, can I work on this issue as my first one? I performed two simple benchmark tests on my laptop. Below you can see, it is a significant difference between power plans.

```csharp
        private readonly Dictionary<string, string> Guids = new Dictionary<string, string>()
        {
            { "HighPerformance", "8c5e7fda-e8bf-4a96-9a85-a6e23a8c635c" },
            { "Balanced", "381b4222-f694-41f0-9685-ff5bb260df2e" },
            { "PowerSaver", "e22fd527-0c09-43ad-83d6-ba300affc27d" }
        };

        [Params("HighPerformance", "Balanced", "PowerSaver")]
        public string PerfPlan { get; set; }

        [IterationSetup]
        public void InitGuid()
        {
            var planGuid = Guids[PerfPlan];
            var guid = new Guid(planGuid);
            PowerManagment.Set(guid);
        }

        [Benchmark]
        public int SplitJoin()
            => string.Join(",", new string[1000]).Split(',').Length;

        [Benchmark]
        public int IterationTest()
        {
            int j = 0;
            for (int i=0; i < short.MaxValue; ++i)
            {
                j = i;
            }

            return j;
        }

```
Method | PerfPlan | Mean | Error | StdDev | Median
-- | -- | -- | -- | -- | --
SplitJoin | Balanced | 32.09 us | 0.8718 us | 2.2348 us | 33.32 us
IterationTest | Balanced | 24.28 us | 0.7690 us | 2.1435 us | 25.38 us
SplitJoin | HighPerformance | 28.58 us | 0.5693 us | 1.0553 us | 27.89 us
IterationTest | HighPerformance | 20.79 us | 0.1932 us | 0.1807 us | 20.68 us
SplitJoin | PowerSaver | 70.12 us | 0.7759 us | 0.7258 us | 69.99 us
IterationTest | PowerSaver | 55.01 us | 0.0000 us | 0.0000 us | 55.01 us

What I suggest that we can use helper class similar to that proposed by @omariom with two methods. The first one will check if the system is windows, save current power plan and eventually set performance to the HighPerf. Then we perform all benchmarks, and after that, we use the second method. The second method will restore the previous user power plan. Both of the methods will execute in the BenchmarkRunner class. What is your opinion about that?

Points to consider:
If someone shutdown system or kill benchmark process, unfortunately, a user will stay with a new performance plan as the side effect of benchmarking. Of course, we can log some information to the user that we changed his power plan.

**Comment by AndreyAkinshin** [comment]
@MarekM25,

> can I work on this issue as my first one? 

Sure!

**Comment by AndreyAkinshin** [comment]
> What I suggest that we can use helper class similar to that proposed by @omariom with two methods. The first one will check if the system is windows, save current power plan and eventually set performance to the HighPerf. Then we perform all benchmarks, and after that, we use the second method. The second method will restore the previous user power plan. Both of the methods will execute in the BenchmarkRunner class. What is your opinion about that?

Looks good to me.
Also, we need an option in Config which allows disabling this behavior.

**Comment by AndreyAkinshin** [comment]
Fixed in #952
